src/main/java/graphql/schema/AsynchronousDataFetcher.java

```diff
@@ -0,0 +1,40 @@
+package graphql.schema;
+
+import static graphql.Assert.assertNotNull;
+
+import java.util.concurrent.CompletableFuture;
+
+import graphql.PublicApi;
+
+/**
+ * A modified type for that indicates the underlying data fetcher is run asynchronously.
+ */
+@PublicApi
+public class AsynchronousDataFetcher<T> implements DataFetcher<CompletableFuture<T>> {
+    
+    /**
+     * A factory method for creating asynchronous data fetchers so that when used with 
+     * static imports allows more readable code such as
+     * {@code .dataFetcher(async(fooDataFetcher)) }
+     *
+     * @param wrappedDataFetcher the data fetcher to run asynchronously
+     *
+     * @return a DataFetcher that will run the wrappedDataFetcher asynchronously
+     */
+    public static <T> AsynchronousDataFetcher<T> async(DataFetcher<T> wrappedDataFetcher) {
+        return new AsynchronousDataFetcher<>(wrappedDataFetcher);
+    }
+    
+    private final DataFetcher<T> wrappedDataFetcher;
+
+    public AsynchronousDataFetcher(DataFetcher<T> wrappedDataFetcher) {
+        assertNotNull(wrappedDataFetcher, "wrappedDataFetcher can't be null");
+        this.wrappedDataFetcher = wrappedDataFetcher;
+    }
+
+    @Override
+    public CompletableFuture<T> get(DataFetchingEnvironment environment) {
+        return CompletableFuture.supplyAsync(() -> wrappedDataFetcher.get(environment));
+    }
```

Comment:
In principle I like aspects of this PR but we cant accept it as it as is because its using a the ForkJoinPool without the ability for some one to specify another ExecutorService.

When ever code uses Threading then the ability to control the underlying ExecutorService is super important.

Managing threadlocals and security managers and so on can be important some some people (it very much is at Atlassian where I work) and hence you need a way to inject  your own ExecutorService into the mix.

For example notice how `graphql.execution.ExecutorServiceExecutionStrategy` takes an executor service say.  This is done for that reason.

Make this code require an ExecutorService, default it to ForkJoinPool.defaultPool() and have the static methods also take an ExecutorService.